cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/SystemAppModule.java

```diff
@@ -0,0 +1,107 @@
+import io.cdap.cdap.app.runtime.ProgramStateWriter;
+import io.cdap.cdap.common.guice.ConfigModule;
+import io.cdap.cdap.common.guice.DFSLocationModule;
+import io.cdap.cdap.common.guice.IOModule;
+import io.cdap.cdap.common.guice.KafkaClientModule;
+import io.cdap.cdap.common.guice.SupplierProviderBridge;
+import io.cdap.cdap.common.guice.ZKClientModule;
+import io.cdap.cdap.common.guice.ZKDiscoveryModule;
+import io.cdap.cdap.common.namespace.guice.NamespaceQueryAdminModule;
+import io.cdap.cdap.data.runtime.DataFabricModules;
+import io.cdap.cdap.data2.metadata.writer.DefaultMetadataServiceClient;
+import io.cdap.cdap.data2.metadata.writer.MetadataServiceClient;
+import io.cdap.cdap.internal.app.program.MessagingProgramStateWriter;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepositoryReader;
+import io.cdap.cdap.internal.app.runtime.artifact.DefaultArtifactRepository;
+import io.cdap.cdap.internal.app.runtime.artifact.RemoteArtifactRepositoryReader;
+import io.cdap.cdap.logging.guice.KafkaLogAppenderModule;
+import io.cdap.cdap.logging.guice.RemoteLogAppenderModule;
+import io.cdap.cdap.master.environment.MasterEnvironments;
+import io.cdap.cdap.master.spi.environment.MasterEnvironment;
+import io.cdap.cdap.messaging.guice.MessagingClientModule;
+import io.cdap.cdap.metrics.guice.MetricsClientRuntimeModule;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.security.auth.context.MasterAuthenticationContext;
+import io.cdap.cdap.security.impersonation.CurrentUGIProvider;
+import io.cdap.cdap.security.impersonation.DefaultImpersonator;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import io.cdap.cdap.security.impersonation.UGIProvider;
+import io.cdap.cdap.security.spi.authentication.AuthenticationContext;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+
+/**
+ * Modules loaded for system app tasks
+ */
+public class SystemAppModule extends AbstractModule {
+
+  @Override
+  protected void configure() {
+    install(new ConfigModule());
+    MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
+
+    if (masterEnv == null) {
+      install(new ZKClientModule());
+      install(new ZKDiscoveryModule());
+      install(new KafkaClientModule());
+      install(new KafkaLogAppenderModule());
+    } else {
+      install(new AbstractModule() {
+        @Override
+        protected void configure() {
+          bind(DiscoveryService.class)
+            .toProvider(new SupplierProviderBridge<>(masterEnv.getDiscoveryServiceSupplier()));
+          bind(DiscoveryServiceClient.class)
+            .toProvider(new SupplierProviderBridge<>(masterEnv.getDiscoveryServiceClientSupplier()));
+        }
+      });
+      install(new RemoteLogAppenderModule());
+    }
+
+    MapBinder.newMapBinder(binder(), ProgramType.class, ProgramRunner.class);
```

Comment:
What is this for? I don't see the MapBinder is being used.

Reply:
This is required for the Map<ProgramType, Provider<ProgramRunner>> injection  in DefaultProgramRunnerFactory.